docs/maintainers/concept-ci.md

```diff
@@ -0,0 +1,125 @@
+# Concept CI
+
+The following tooling can be used to keep concept names consistent within a track.
+
+## `concepts.csv`
+
+This machine-readable file defines all concept slugs used in the track.
+The first column contains the concept.
+The second column **optionally** contains the category of the concept, e.g. some tracks differentiate between functional and object-oriented concepts.
+The first row contains the headers, i.e. `concept,category`.
+
+Concepts may contain annotations separated with a dot, `.`, from the concept name, e.g. `numbers.basic` and `numbers.advanced`.
+These annotations can be used when a single concept has multiple exercises.
+As a rule of thumb, annotations should only be used when multiple exercises refer to the same concept document, e.g. `string-formatting` might have its own document, while `strings.basic` and `strings.advanced` both refer to `strings.md`.
+
+### Example
+
+```csv
+concept,category
+encapsulation,object-oriented
+classes,object-oriented
+inheritance,object-oriented
+anonymous-functions,functions
+higher-order-functions,functions
+local-functions,functions
+numbers.basic,
+numbers.advanced,
+```
+
+For a longer example, see the Julia track's [`concepts.csv`][julia-concepts-csv].
+
+## CI checks
+
+*If any issues with the CI script come up, please ping @SaschaMann in the PR/issue.*
+
+The following checks are currently available:
+
+- Check if all concepts used by exercises in `config.json` are defined in `concepts.csv`
+  - **sensitive to annotations** (e.g. `strings.basic` in concepts.csv will be treated as `strings.basic`. if `concepts.csv` only contains `strings`, it will error)
+- Check if concept extraction docs only contains valid concepts, under the following assumptions:
+  - The first list is ignored, as it is assumed to contain the example implementations
+  - In the rest of the doc, all lists are assumed to have the following format: `- <concept>: <explanation>`. Entries that aren't lists are ignored
+  - **sensitive to annotations**
+- Check if all exercise directory names are valid concepts.
+  - **sensitive to annotations**
+
+### Workflow file
+
+You can install the Concept CI checks for your track by creating a workflow in `.github/workflows/<language>-concept-ci.yml` with the following contents:
```

Comment:
I'm wondering if we could instead have the tooling detect if a track has a `concepts.csv` file, and if not, just skip the run. If we do that, any track that would add the `concepts.csv` file would automatically be included in the CI checks.

Reply:
I only want this to run on the track that the push/PR modifies.

I could think of two ways to accomplish that:

1. One workflow per track.
2. One workflow with a long `paths:` section that includes all relevant files in all participating tracks, detects which track is affected by the PR and then only runs the checks on that.

While it has the downside of creating one workflow file per track, I went with 1) for two reasons. It's simpler to setup and, more importantly, it's simpler to customise the CI checks if a track only wants to run some of them. The latter is not supported by the script yet, but it would be fairly easy to add. For example, it might be possible that some tracks use a different format for their concept extraction exercises, but still want to run all the other checks.

Reply:
I see. Let me think on this a bit.

Reply:
Due to the added complexity of the workflow file introduced in https://github.com/exercism/v3/pull/718/commits/85be4b76e7ed2b193361efa9a091efeca457bdf6, it might be best to go with 2) and have one workflow file that gets triggered by changes to

```yaml
    - 'languages/julia/config.json'
    - 'languages/julia/reference/concepts.csv'
    - 'languages/julia/reference/exercise-concepts/**'
    - 'languages/julia/exercises/concept/**'
    - 'languages/rust/config.json'
    - 'languages/rust/reference/concepts.csv'
    - 'languages/rust/reference/exercise-concepts/**'
    - 'languages/rust/exercises/concept/**'
    - ...
```

and then detects which track the PR changes.

Reply:
I'll fix this another day

Reply:
Great!

Reply:
Done!